### PR TITLE
Tips on Building Drivers for Zephyr post

### PR DESCRIPTION
**Pull request by jaredwolff**
Hey @tyhoff, here's my post. I've yet to proofread it fully. I'll give it another good scrubbing in a bit. Let me know what you think!

### CODE REVIEW

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+`west` is the de facto tool for Zephyr projects. It makes project management, compilation and flashing easier. It was built to replace the idea of a top level Makefile and project dependency management like Git submodules. For the most part, it accomplishes that goal.
+
+You'll definitely need to install `west` if you haven't already. Follow the SDK setup links above for more information.
+
+**Building**
+
+To build a project you'll either have to change directories to the application in question or provide the path. For example:
+
+```terminal
+cd ncs/nrf/samples/nrf9160/nrf_cloud_agps
```
Comment (tyhoff) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:38]: nit: prefix shell commands with $ for readability

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+
+### `CMakeLists.txt`
+
+If you look carefully, many folders and sub-folders in Zephyr have a consistent pair of files: `CMakeLists.txt` and `Kconfig`. These files is where all the magic happens.
+
+You may already have an idea of what `CMakeLists.txt` does by its name. It's a Makefile of sorts that is used to add source files, folders and more to a project. The neat thing about `cmake` is that it's recursive (for better or for worse) and higher level `CMakeLists.txt` files dictate what, if anything, is below it. There's no more ginormous Makefiles. You make the changes you need closest to the code that's affected.
+
+Sometimes `CMakeLists.txt` are used to simply "include" more directories like in this example:
+
+```
```
Comment (tyhoff) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:107]: Should be able to highlight CMake with using

"```cmake"


at the top. For the list of other options, you can check here: http://rouge.jneen.net/

**Comment by jaredwolff** [comment]
This is awesome. Thanks for the feedback @tyhoff. I'll get back to it this morning. :)

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 1 above)
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:38]: Adjusted

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 2 above)
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md]: Fixed! I didn't see one for Kconfig though. Bummer.

**Comment by jaredwolff** [comment]
I've given it a full scrub. Grammarly + Hemmingway. Revamped the intro. Let me know what you think @tyhoff

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+## Adding a drivers folder
+
+You now should understand `Kconfig`, `CMakeLists.txt`, and `module.yml`. We're ready for driver development!
+
+ In my case, since my driver work is ongoing, I opted to use my own `drivers` folder in my main project repository. You can always opt to start developing your drivers in `zephyr/drivers`. This is likely a better location especially if you want to release the driver for others to use. The downside is that it requires more work!
+
+In my case, here's what the directory structure looks like from my base repo:
+
+```terminal
+.
```
Comment (chrisc11) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:209]: Might call out the root directory location here too, just to be extra explicit
```suggestion
. # My Project Root, pyrinas
```

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+
+A ton of things happen in the background to build a project. We'll discuss more how things compile in the next section.
+
+**Flashing**
+
+Flashing code is as simple as running `west flash` from your project directory. You can use the `-r` parameter to choose which programming method you'd like to use. Common options are `jlink`, `nrfjprog` (for Nordic projects), `pyocd`, `openocd`, `dfu-util`, etc.
+
+These "runners" are set in the board configuration. Here's the `board.cmake` for the nRF9160 Feather:
+
+```cmake
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:58]: I'll add both!

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+
+### What `west` can do for you
+
+`west` It makes project management, compilation, and flashing easier. `west` replaces the idea of a top-level Makefile and project dependency management like Git submodules. For the most part, it accomplishes that goal.
+
+You’ll need to install `west` if you haven't already. Follow the SDK setup links above for more information.
+
+**Building**
+
+To build a project you have to change directories to the application in question or provide the path. For example:
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:41]: Changed!

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+
+There are some prerequisites to getting started. Let's begin with SDK Setup.
+
+### SDK Setup
+
+I'll be referencing Nordic's nRF Connect SDK a bunch throughout this post. I've been building and testing nRF9160 Feather on v1.3.0. I won't be going into detail on how to set up the SDK though. Nordic has [great documentation](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/getting_started.html) you can check out to get started. I've also discussed [my experience here](https://www.jaredwolff.com/zephyr-on-nrf9160-feather/).
+
+### What `west` can do for you
+
+`west` It makes project management, compilation, and flashing easier. `west` replaces the idea of a top-level Makefile and project dependency management like Git submodules. For the most part, it accomplishes that goal.
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:35]: Changed!

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 7]
```diff
@@ -0,0 +1,576 @@
...
+
+## How Things Compile
+
+One of the important things to know about driver development is how things compile. It may bend your mind a bit but we'll get through this together.
+
+### `CMakeLists.txt`
+
+If you've spent enough time in Zephyr, you may see some patterns emerge. Many folders and sub-folders have this dynamic duo: `CMakeLists.txt` and `Kconfig`. These files are where all the magic happens.
+
+You may already have an idea of what `CMakeLists.txt` does by its name. It's a Makefile of sorts that you use to add source files and include folders to a project. The neat thing about `cmake` is that it's recursive (for better or for worse). This means higher-level `CMakeLists.txt` files dictate what, if anything, is below it. There are no more ginormous Makefiles. You make the changes you need closest to the code that's affected.
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:107]: Changed

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+
+As with any device, your driver may look different. The Zephyr project is full of example driver implementations. You can check out more in the `nrf/drivers/` folder and also the `zephyr/drivers` folder.
+
+## Driver Smarts
+
+Now that you're configured and your Device Tree is set, it's time to write some code! In this section, we'll review the important macros and nuances of driver development.
+
+### Helpers
+
+Who doesn't like a handy macro that makes life easier? I found that the macros like `BIT` and `BIT_MASK` were extra useful. Here's what it looks like for the CTRL1 register of the PCF85063A
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:374]: `zephyr/include/sys/util.h` 

I'll add that bit in 😎

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 9]
```diff
@@ -0,0 +1,576 @@
...
+
+Though these macros don't do too much, they keep your code clean and easy to read.
+
+### `DEVICE_AND_API_INIT`
+
+The `DEVICE_AND_API_INIT` macro adds the initialization of your driver before any user code. Here's what the setup looks like for the PCF85063A:
+
+```c
+DEVICE_AND_API_INIT(pcf85063a, DT_INST_LABEL(0), pcf85063a_init, &pcf85063a_data,
+										&pcf85063_cfg_info, POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:397]: Fixed

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 10]
```diff
@@ -0,0 +1,589 @@
...
+As an example, here's what the `ncs/nrf/drivers/sensor/CMakeLists.txt` looks like:
+
+```cmake
+add_subdirectory_ifdef(CONFIG_BH1749 bh1749)
+add_subdirectory_ifdef(CONFIG_SENSOR_SIM sensor_sim)
+add_subdirectory_ifdef(CONFIG_PMW3360 pmw3360)
+add_subdirectory_ifdef(CONFIG_PAW3212 paw3212)
+```
+
+You may noice that all the variables begin with `CONFIG_`. All variables produced by `Kconfig` have this prefix. So, for example, `APPLICATION_WORKQUEUE_STACK_SIZE` becomes `CONFIG_APPLICATION_WORKQUEUE_STACK_SIZE.`
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:181]: Fixed

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 11]
```diff
@@ -0,0 +1,589 @@
...
+static const struct counter_driver_api pcf85063a_driver_api = {
+		.start = pcf85063a_start,
+		.stop = pcf85063a_stop,
+		.get_value = pcf85063a_get_value,
+		.set_alarm = pcf85063a_set_alarm,
+		.cancel_alarm = pcf85063a_cancel_alarm,
+		.set_top_value = pcf85063a_set_top_value,
+		.get_pending_int = pcf85063a_get_pending_int,
+		.get_top_value = pcf85063a_get_top_value,
+		.get_max_relative_alarm = pcf85063a_get_max_relative_alarm,
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:458]: Removed those pesky indents.

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 3 above)
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md]: Ahh good call. Fixed that!

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 12]
```diff
@@ -0,0 +1,591 @@
...
+	pcf85063a@51 {
+			compatible = "nxp,pcf85063a";
+			label = "PCF85063A";
+			reg = <0x51>;
+	};
+
+};
+```
+
+I got my inspiration from the drivers for the `bh1749` driver in the nRF Connect SDK. You can see how Nordic implemented it in `thingy91_nrf910_common.dts`.
```
Comment (tyhoff) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:360]: Link to thingy91_nrf910_common in Github maybe?

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 13]
```diff
@@ -0,0 +1,591 @@
...
+static const struct counter_driver_api pcf85063a_driver_api = {
+.start = pcf85063a_start,
+.stop = pcf85063a_stop,
+.get_value = pcf85063a_get_value,
+.set_alarm = pcf85063a_set_alarm,
+.cancel_alarm = pcf85063a_cancel_alarm,
+.set_top_value = pcf85063a_set_top_value,
+.get_pending_int = pcf85063a_get_pending_int,
+.get_top_value = pcf85063a_get_top_value,
+.get_max_relative_alarm = pcf85063a_get_max_relative_alarm,
```
Comment (tyhoff) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:459]:
```suggestion
static const struct counter_driver_api pcf85063a_driver_api = {
    .start = pcf85063a_start,
    .stop = pcf85063a_stop,
    .get_value = pcf85063a_get_value,
    .set_alarm = pcf85063a_set_alarm,
    .cancel_alarm = pcf85063a_cancel_alarm,
    .set_top_value = pcf85063a_set_top_value,
    .get_pending_int = pcf85063a_get_pending_int,
    .get_top_value = pcf85063a_get_top_value,
    .get_max_relative_alarm = pcf85063a_get_max_relative_alarm,
```

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 14]
```diff
@@ -0,0 +1,591 @@
+---
+title: "Tips on Building Drivers for Zephyr"
```
Comment (tyhoff) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:2]: Maybe: Guide to Writing New Drivers for Zephyr

Your post is more than just tips...it's pretty much a guide

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+description: "A guide on how to build custom drivers for hardware on Zephyr."
+author: jaredwolff
+---
```
Comment (tyhoff) [inline _posts/2020-07-16-building-drivers-on-zephyr.md:5]: The cover photo will default to the Interrupt one if another is not given. https://interrupt.memfault.com/blog/img/interrupt-cover-1200px.png. Feel free to add a custom one if you like!

**Comment by jaredwolff** [comment]
Thanks @tyhoff. I'll get to fixing these bits and we'll go from there.

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 14 above)
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:2]: changed it to _How to Build Drivers for Zephyr_ How-to's always do well. 👍

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 16]
```diff
@@ -0,0 +1,591 @@
...
+You now should understand `Kconfig`, `CMakeLists.txt`, and `module.yml`. We're ready for driver development!
+
+ In my case, since my driver work is ongoing, I opted to use my own `drivers` folder in my main project repository. You can always opt to start developing your drivers in `zephyr/drivers`. This is likely a better location especially if you want to release the driver for others to use. The downside is that it requires more work!
+
+In my case, here's what the directory structure looks like from my project repo:
+
+```terminal
+. # My Project Root located in /ncs/pyrinas/
+├── CMakeLists.txt
+├── Kconfig.pyrinas
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:212]: Good question. I was following the convention in the other `module.yml` files in Zephyr and NCS. Their Kconfig files end with `.nrf` and `.zephyr`. Is it truly necessary? I don't think so unless you need a raw `Kconfig` file in the root as well for some reason.

File: _posts/2020-07-16-building-drivers-on-zephyr.md [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Setup + Tools
+
+There are some prerequisites to getting started. Let's begin with SDK Setup.
```
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:27]: Ahh maybe something like this:

_In this example, you'll learn about what it takes to create a driver for the PCF85063A real-time clock and counter. I've chosen to use the counter API for reasons we'll discuss in a bit. There are some prerequisites to getting started though. Let's begin with SDK Setup._

[the remark this answers is not in the document]
Comment (tyhoff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md]: 👏 👏 👏

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (tyhoff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:213]: I'll leave it up to you then :D. I just got confused when I saw it, that's all.

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 15 above)
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:6]: Done.

File: _posts/2020-07-16-building-drivers-on-zephyr.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jaredwolff) [reply _posts/2020-07-16-building-drivers-on-zephyr.md:214]: Ahh. I added a side note about this a little later on

**Comment by tyhoff** [comment]
Looks great! Unless you have anything else, I'll merge this later tonight.

**Comment by tyhoff** [comment]
Going to merge this into the branch `staging` and prep it for release. Just a few things to do that I don't want to burden you with.

**Comment by jaredwolff** [comment]
Cool. Thanks @tyhoff 

My launch is a little delayed to some last minute administrative issues. I'd love to have this article in sync with the launch if possible. 😄

I'm thinking everything will go live tomorrow but in the case it goes live Thursday, can we hold off a day?

**Comment by tyhoff** [comment]
@jaredwolff Oh! I didn't realize there was a coordinated launch. Of course! Keep me updated (email is best). I can do the push when you're ready.

**Comment by jaredwolff** [comment]
Yea I didn't communicate that very well! I'll shoot you an email when things are good to go. Should be tomorrow but these things never go as planned.

**Comment by jaredwolff** [comment]
@tyhoff good to go for tomorrow mid-day PDT. I'll email as you as well.

**Comment by jaredwolff** [comment]
@tyhoff looks like the image at the bottom of the post barfed. Must have been something funky from the Jekyll compile. Can you take a peek at your earliest convenience?

**Comment by tyhoff** [comment]
@jaredwolff Pushed a fix. Thanks for letting me know.

**Pull request merged**